### ISSUE TITLE
Add explicit ability to choose order of Tapioca compiler execution

### ISSUE THREAD

**Issue by bitwise-aiden** [Shopify/tapioca#1965]
### Description 

At present when compilers for DSL generation are loaded they are alphabetized and run in that order:

https://github.com/Shopify/tapioca/blob/8678e2ab7b853ccbf3f092ccf553e58773eaa38b/lib/tapioca/dsl/pipeline.rb#L115-L123

What this effectively means is that if two compilers create the same method with different signatures, whichever compiler was run last will be the one to define the signature used. If someone were wanting to write a different signature than a Tapioca defined one, their only recourse would be to name their compiler something that would come after `Tapioca::Dsl::Compilers::*` alphabetically.

Having the ability to push the compiler forward or back in the run order would make it possible to have the explicit control over the order rather than relying on discovering the above code to then know to name it something that puts it in the right place. Naming is already hard enough, but name golf to get your compiler to run in order although fun seems a little counter intuitive.

#### Order value

One solution would be to have a `order` value which is an integer representing where it is expected to run. This could then update the code above to order based on this value first, then alphabetically next to ensure a stable order run. The order value could be an integer between 0-100 and by default if not set, the could be 50 to allow for room either side. 

This would then allow projects to set the value of their compilers based on what they are needed and even override ones from their gems/Tapioca to ensure the order needed for them.

#### Method override flags

Given that the heart of the problem is that a method needs to be defined after an existing to override it's signature, maybe when generating it, optional flags can be passed to determine how it interacts with existing / future methods of the same name. It would have to be able to handle future cases too because of the aforementioned ordering alphabetically.

**Comment by bitwise-aiden** [comment]
> For the different signature use case is it enough to run custom compilers after default ones?

Yes, I think providing the way to run them after would likely hit 99+% of use cases. Separating them out would hopefully fix that.

**Comment by paracycle** [comment]
I'd like to question the premise of the issue, though. What is a real life example where 2 compilers want to type the same method in different ways?

I'd be wary of solutionizing before we understand if the problem is really a problem.

**Comment by paracycle** [comment]
This should allow you to solve that: https://github.com/Shopify/tapioca/pull/1967

**Comment by paracycle** [comment]
I also added a DSL CLI test that shows how to use the extensibility point inside a custom "compiler".

**Comment by bitwise-aiden** [comment]
Although it fixes this one example, I do think that there is very likely a need for others to override behaviours. I think the layered `tapioca > gems > project` ordering is probably a good quality of life feature so fixes don't have to be upstreamed.

**Comment by paracycle** [comment]
> Although it fixes this one example, I do think that there is very likely a need for others to override behaviours. I think the layered `tapioca > gems > project` ordering is probably a good quality of life feature so fixes don't have to be upstreamed.

None of those should require a generation time modification. All modifications should be doable at load time when compilers are being loaded.

**Comment by paracycle** [comment]
Btw, for the `find` signature, I am actually pretty unhappy with the fact that we've limited the set of id types artificially, when it is obvious that that isn't enough. So, I generally think this is a short-coming of the compiler itself.

I remember that we had a problem with overloads not working as expected when one of them was `T.untyped` but I think if we are careful about the overload sig ordering, we can still use `T.untyped`:
```
# typed: __STDLIB_INTERNAL

class Foo
    extend T::Sig

    sig { params(a: T::Array[T.untyped]).returns(Integer) }
    sig { params(a: T.untyped).returns(String) }
    def self.foo(a)
      T.unsafe(nil)
    end
end

T.reveal_type Foo.foo(1) # Revealed type: String
T.reveal_type Foo.foo([1, "a"]) # Revealed type: Integer
T.reveal_type Foo.foo([[1, "a"]]) # Revealed type: Integer
```
[seems to work just fine](https://sorbet.run/?arg=--enable-experimental-requires-ancestor#%23%20typed%3A%20__STDLIB_INTERNAL%0A%0Aclass%20Foo%0A%20%20%20%20extend%20T%3A%3ASig%0A%0A%20%20%20%20sig%20%7B%20params%28a%3A%20T%3A%3AArray%5BT.untyped%5D%29.returns%28Integer%29%20%7D%0A%20%20%20%20sig%20%7B%20params%28a%3A%20T.untyped%29.returns%28String%29%20%7D%0A%20%20%20%20def%20self.foo%28a%29%0A%20%20%20%20%20%20T.unsafe%28nil%29%0A%20%20%20%20end%0Aend%0A%0AT.reveal_type%20Foo.foo%281%29%0AT.reveal_type%20Foo.foo%28%5B1%2C%20%22a%22%5D%29%0AT.reveal_type%20Foo.foo%28%5B%5B1%2C%20%22a%22%5D%5D%29) and is a more "correct" signature for the method.

@KaanOzkan do you remember why we went the way of specifying all these id types?

**Comment by paracycle** [comment]
Ah, found the problem: https://github.com/Shopify/tapioca/pull/1799#discussion_r1511362645
```ruby
# typed: __STDLIB_INTERNAL

class Foo
    extend T::Sig

    sig { params(a: T::Array[T.untyped]).returns(Integer) }
    sig { params(a: T.untyped).returns(String) }
    def self.foo(a)
      T.unsafe(nil)
    end
end

array_of_untyped = T.let(T.unsafe(nil), T::Array[T.untyped])
untyped = T.let(T.unsafe(nil), T.untyped)

T.reveal_type Foo.foo(untyped) # Revealed type: Integer
T.reveal_type Foo.foo(array_of_untyped) # Revealed type: Integer
```
[which is really strange](https://sorbet.run/?arg=--enable-experimental-requires-ancestor#%23%20typed%3A%20__STDLIB_INTERNAL%0A%0Aclass%20Foo%0A%20%20%20%20extend%20T%3A%3ASig%0A%0A%20%20%20%20sig%20%7B%20params%28a%3A%20T%3A%3AArray%5BT.untyped%5D%29.returns%28Integer%29%20%7D%0A%20%20%20%20sig%20%7B%20params%28a%3A%20T.untyped%29.returns%28String%29%20%7D%0A%20%20%20%20def%20self.foo%28a%29%0A%20%20%20%20%20%20T.unsafe%28nil%29%0A%20%20%20%20end%0Aend%0A%0Aarray_of_untyped%20%3D%20T.let%28T.unsafe%28nil%29%2C%20T%3A%3AArray%5BT.untyped%5D%29%0Auntyped%20%3D%20T.let%28T.unsafe%28nil%29%2C%20T.untyped%29%0A%0AT.reveal_type%20Foo.foo%28untyped%29%0AT.reveal_type%20Foo.foo%28array_of_untyped%29%0A).

**Comment by bitwise-aiden** [comment]
> None of those should require RBI generation time modification. All modifications should be doable at load time when compilers are being loaded.

Exactly and that is the issue that I'm having, we load the compilers, order them alphabetically, then run them. What I'm proposing is that we load them, order them by context AND alphabetically, then run them.

Something like:
```ruby
TAPIOCA = 0
GEMS = 1
LOCAL = 2

Runtime::Reflection.descendants_of(Compiler).sort_by do |compiler|
  context = context_for_compiler(compiler)
  [context, T.must(compiler.name]
end,

> What happens if you change the order of the signatures?

The main issue I was having is that two compilers generate two separate methods and can't impact the order of the sigs, so whichever one is read last will be the one that's sigs are applied.
e.g.

```ruby
class SomeModel
  module CommonRelationMethods
    sig { params(args: SomeIdType).returns(SomeModel)
    def find(args = nil, &block); end
    
    sig { params(args: T.untyped).returns(T.untyped)
    def find(args = nil, &block); end
  end
end
```

The sig with `SomeIdType` will never be used.

**Comment by paracycle** [comment]
> Exactly and that is the issue that I'm having, we load the compilers, order them alphabetically, then run them.

What I was trying to say, but couldn't word properly, was that the compiler run order should be immaterial for properly constructed DSL compilers. That's why I am slightly opposed to implementing a certain kind of run order for compilers.

The main behaviour of DSL compilers should have been settled at load time, before we get to running. When we run the compilers, their behaviours should not depend on each other or on any runtime ordering (which implies the same thing).

**Comment by paracycle** [comment]
> I think there are a rich set of additional tooling that would help to create signatures across compilers, but with what we have I think there is a bit of a short coming.

That might be true, but I think we should address that as a problem on its own and not as a problem of DSL compiler ordering.

For example, we could create a `create_or_modify_method` helper for DSL compilers which they could use that exclusively and has some kind of conflict resolution semantics built in. That might be another way for DSL compiler to work collaboratively.

Basically, I'd like to keep the solution space wide enough to allow those kinds of solutions without committing to any ordering guarantees.

**Comment by bitwise-aiden** [comment]
> For example, we could create a create_or_modify_method helper for DSL compilers which they could use that exclusively and has some kind of conflict resolution semantics built in. That might be another way for DSL compiler to work collaboratively.

+1 for this. I was definitely wanting for this over the weekend.

**Comment by paracycle** [comment]
I am closing this issue, since the specific ask is, in my opinion, not something we want and the root cause for needing this was addressed in #1967. 

If you would like to open another issue with concrete examples of DSL compilers needing cross-compiler collaboration, we could discuss the possible approaches in there.
